File: app/src/main/java/com/veganbeauty/admin/features/customer/CustomerCreateCampaignFragment.java
package com.veganbeauty.admin.features.customer;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.veganbeauty.admin.MainActivity;
import com.veganbeauty.admin.R;
import com.veganbeauty.admin.core.base.RootieAdminFragment;
import com.veganbeauty.admin.databinding.CustomerCreateCampaignBinding;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Random;

public class CustomerCreateCampaignFragment extends RootieAdminFragment {

    private CustomerCreateCampaignBinding binding;
    private final List<String> categoriesList = new ArrayList<>();
    private final List<String> offerTypes = Arrays.asList("Giảm giá phần trăm (%)", "Giảm giá cố định (VNĐ)");

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = CustomerCreateCampaignBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    protected void setupUI(View view) {
        setupListeners();
        loadCategories();
        setupOfferTypeSpinner();
        setupInputWatchers();
        updateSaveButtonState();
    }

    private void setupListeners() {
        binding.btnBack.setOnClickListener(v -> navigateBack());
        binding.btnCancel.setOnClickListener(v -> navigateBack());

        binding.edtStartDate.setOnClickListener(v -> showDatePicker(binding.edtStartDate));
        binding.edtEndDate.setOnClickListener(v -> showDatePicker(binding.edtEndDate));

        binding.btnSaveCampaign.setOnClickListener(v -> saveCampaign());
    }

    private void setupInputWatchers() {
        TextWatcher watcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                updateSaveButtonState();
            }
            @Override
            public void afterTextChanged(Editable s) {}
        };

        binding.edtCampaignName.addTextChangedListener(watcher);
        binding.edtCampaignDesc.addTextChangedListener(watcher);
        binding.edtMinOrderValue.addTextChangedListener(watcher);
        binding.edtDiscountValue.addTextChangedListener(watcher);
        binding.edtStartDate.addTextChangedListener(watcher);
        binding.edtEndDate.addTextChangedListener(watcher);
    }

    private void updateSaveButtonState() {
        String title = binding.edtCampaignName.getText().toString().trim();
        String description = binding.edtCampaignDesc.getText().toString().trim();
        String minOrder = binding.edtMinOrderValue.getText().toString().trim();
        String discount = binding.edtDiscountValue.getText().toString().trim();
        String startDate = binding.edtStartDate.getText().toString().trim();
        String endDate = binding.edtEndDate.getText().toString().trim();

        boolean allFilled = !title.isEmpty() &&
                            !description.isEmpty() &&
                            !minOrder.isEmpty() &&
                            !discount.isEmpty() &&
                            !startDate.isEmpty() &&
                            !endDate.isEmpty();

        binding.btnSaveCampaign.setEnabled(allFilled);
        binding.btnSaveCampaign.setAlpha(allFilled ? 1.0f : 0.5f);
    }

    private void showDatePicker(EditText editText) {
        Calendar calendar = Calendar.getInstance();
        String currentText = editText.getText().toString();
        if (!currentText.isEmpty()) {
            try {
                Date date = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).parse(currentText);
                if (date != null) {
                    calendar.setTime(date);
                }
            } catch (Exception ignored) {
            }
        }

        DatePickerDialog datePickerDialog = new DatePickerDialog(
            requireContext(),
            (view, year, month, dayOfMonth) -> {
                Calendar selectedCal = Calendar.getInstance();
                selectedCal.set(Calendar.YEAR, year);
                selectedCal.set(Calendar.MONTH, month);
                selectedCal.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault());
                editText.setText(sdf.format(selectedCal.getTime()));
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        );
        datePickerDialog.show();
    }

    private void loadCategories() {
        if (getActivity() == null) return;
        new Thread(() -> {
            List<String> loadedCategories = new ArrayList<>();
            loadedCategories.add("Tất cả sản phẩm"); // Default option
            try {
                StringBuilder sb = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(requireContext().getAssets().open("categories.json")))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        sb.append(line);
                    }
                }
                JSONObject jsonObject = new JSONObject(sb.toString());
                JSONArray jsonArray = jsonObject.getJSONArray("categories");
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject catObj = jsonArray.getJSONObject(i);
                    String name = catObj.optString("name", "");
                    if (!name.isEmpty() && !loadedCategories.contains(name)) {
                        loadedCategories.add(name);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }

            if (getActivity() != null) {
                getActivity().runOnUiThread(() -> {
                    categoriesList.clear();
                    categoriesList.addAll(loadedCategories);

                    ArrayAdapter<String> adapter = new ArrayAdapter<>(
                        requireContext(),
                        android.R.layout.simple_spinner_item,
                        categoriesList
                    );
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                    binding.spnApplicableProducts.setAdapter(adapter);
                });
            }
        }).start();
    }

    private void setupOfferTypeSpinner() {
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
            requireContext(),
            android.R.layout.simple_spinner_item,
            offerTypes
        );
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        binding.spnOfferType.setAdapter(adapter);

        binding.spnOfferType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position == 0) {
                    binding.txtDiscountLabel.setText("Giá trị (%)");
                    binding.edtDiscountValue.setHint("10");
                } else {
                    binding.txtDiscountLabel.setText("Giá trị (VNĐ)");
                    binding.edtDiscountValue.setHint("50000");
                }
            }
            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    private void saveCampaign() {
        String title = binding.edtCampaignName.getText().toString().trim();
        String description = binding.edtCampaignDesc.getText().toString().trim();
        String minOrderStr = binding.edtMinOrderValue.getText().toString().trim();
        String discountValStr = binding.edtDiscountValue.getText().toString().trim();
        String startDateStr = binding.edtStartDate.getText().toString().trim();
        String endDateStr = binding.edtEndDate.getText().toString().trim();

        if (title.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng nhập tên chương trình!", Toast.LENGTH_SHORT).show();
            return;
        }

        if (description.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng nhập mô tả chương trình!", Toast.LENGTH_SHORT).show();
            return;
        }

        long minOrderValue = 0L;
        try {
            minOrderValue = Long.parseLong(minOrderStr);
        } catch (NumberFormatException ignored) {}

        Long discountValue = null;
        try {
            discountValue = Long.parseLong(discountValStr);
        } catch (NumberFormatException ignored) {}

        if (discountValue == null || discountValue <= 0) {
            Toast.makeText(requireContext(), "Vui lòng nhập giá trị khuyến mãi hợp lệ!", Toast.LENGTH_SHORT).show();
            return;
        }

        if (endDateStr.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng chọn ngày kết thúc!", Toast.LENGTH_SHORT).show();
            return;
        }

        final long finalMinOrderValue = minOrderValue;
        final long finalDiscountValue = discountValue;

        if (getActivity() == null) return;
        new Thread(() -> {
            try {
                // 1. Get writeable file in internal storage
                File internalFile = new File(requireContext().getFilesDir(), "vouchers.json");

                // 2. Load existing vouchers
                JSONArray jsonArray;
                if (internalFile.exists()) {
                    StringBuilder sb = new StringBuilder();
                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(internalFile)))) {
                        String line;
                        while ((line = reader.readLine()) != null) {
                            sb.append(line);
                        }
                    }
                    jsonArray = new JSONArray(sb.toString());
                } else {
                    StringBuilder sb = new StringBuilder();
                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(requireContext().getAssets().open("vouchers.json")))) {
                        String line;
                        while ((line = reader.readLine()) != null) {
                            sb.append(line);
                        }
                    }
                    jsonArray = new JSONArray(sb.toString());
                }

                // 3. Generate unique ID & Code
                int maxVal = 0;
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject item = jsonArray.getJSONObject(i);
                    String idStr = item.optString("id", "");
                    if (idStr.startsWith("v")) {
                        try {
                            int num = Integer.parseInt(idStr.substring(1));
                            if (num > maxVal) maxVal = num;
                        } catch (NumberFormatException ignored) {}
                    }
                }
                int nextIdNum = maxVal + 1;
                String newId = "v" + String.format(Locale.getDefault(), "%03d", nextIdNum);

                // Voucher Type Mapping
                String voucherType = binding.radTypeDiscount.isChecked() ? "discount" : "free ship";

                // Offer Type Mapping
                String offerType = binding.spnOfferType.getSelectedItemPosition() == 0 ? "percentage" : "fixed_amount";

                // Auto-generated Voucher Code
                String autoCode = generateVoucherCode(title, offerType, finalDiscountValue);

                // Format HSD: end date with 23:59:59 time
                String hsdVal;
                try {
                    Date date = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).parse(endDateStr);
                    if (date != null) {
                        hsdVal = new SimpleDateFormat("yyyy-MM-year", Locale.getDefault()).format(date)
                            .replace("year", new SimpleDateFormat("yyyy", Locale.getDefault()).format(date)) + " 23:59:59";
                        // Note: To avoid year format complications in simpledateformat, let's just do:
                        hsdVal = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(date) + " 23:59:59";
                    } else {
                        hsdVal = "2026-12-31 23:59:59";
                    }
                } catch (Exception e) {
                    hsdVal = "2026-12-31 23:59:59";
                }

                // Selected applicable category name
                String selectedCategory = binding.spnApplicableProducts.getSelectedItem() != null ? binding.spnApplicableProducts.getSelectedItem().toString() : "Tất cả sản phẩm";

                // 4. Create new voucher JSON object
                JSONObject newVoucher = new JSONObject();
                newVoucher.put("id", newId);
                newVoucher.put("title", title);
                newVoucher.put("description", description);
                newVoucher.put("code", autoCode);
                newVoucher.put("status", "valid");
                newVoucher.put("hsd", hsdVal);
                newVoucher.put("type", voucherType);
                newVoucher.put("from-gift", false);
                newVoucher.put("quantity", 50); // Default quantity
                newVoucher.put("minOrderValue", finalMinOrderValue);
                newVoucher.put("applicableProducts", selectedCategory);
                newVoucher.put("offerType", offerType);
                newVoucher.put("discountValue", finalDiscountValue);

                // 5. Append & Save
                jsonArray.put(newVoucher);
                try (FileOutputStream fos = new FileOutputStream(internalFile)) {
                    fos.write(jsonArray.toString(2).getBytes());
                }

                if (getActivity() != null) {
                    getActivity().runOnUiThread(() -> {
                        Toast.makeText(requireContext(), "Tạo chương trình khuyến mãi thành công!", Toast.LENGTH_SHORT).show();
                        navigateBack();
                    });
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    private String generateVoucherCode(String campaignName, String offerType, long discountValue) {
        String[] words = campaignName.split(" ");
        StringBuilder prefixBuilder = new StringBuilder();
        int count = 0;
        for (String w : words) {
            if (!w.trim().isEmpty() && count < 3) {
                char firstChar = w.trim().charAt(0);
                if (Character.isLetter(firstChar)) {
                    prefixBuilder.append(Character.toUpperCase(firstChar));
                    count++;
                }
            }
        }
        String prefix = prefixBuilder.toString();
        String cleanPrefix = prefix.isEmpty() ? "KM" : prefix;
        String suffix = offerType.equals("percentage") ? discountValue + "PCT" : (discountValue / 1000) + "K";

        String charPool = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
        StringBuilder randomPart = new StringBuilder();
        Random random = new Random();
        for (int i = 0; i < 3; i++) {
            randomPart.append(charPool.charAt(random.nextInt(charPool.length())));
        }

        String combined = cleanPrefix + suffix + randomPart.toString();
        if (combined.length() > 12) {
            combined = combined.substring(0, 12);
        }
        return combined.toUpperCase(Locale.getDefault());
    }

    private void navigateBack() {
        MainActivity mainActivity = (MainActivity) getActivity();
        if (mainActivity != null) {
            mainActivity.loadFragment(new CustomerAdminFragment());
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/admin/features/customer/CustomerCreateCampaignFragment.kt
package com.veganbeauty.admin.features.customer

import android.app.DatePickerDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import com.veganbeauty.admin.MainActivity
import com.veganbeauty.admin.R
import com.veganbeauty.admin.core.base.RootieAdminFragment
import com.veganbeauty.admin.databinding.CustomerCreateCampaignBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStreamReader
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class CustomerCreateCampaignFragment : RootieAdminFragment() {

    private var _binding: CustomerCreateCampaignBinding? = null
    private val binding get() = _binding!!

    private val categoriesList = mutableListOf<String>()
    private val offerTypes = listOf("Giảm giá phần trăm (%)", "Giảm giá cố định (VNĐ)")

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = CustomerCreateCampaignBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun setupUI(view: View) {
        setupListeners()
        loadCategories()
        setupOfferTypeSpinner()
        setupInputWatchers()
        updateSaveButtonState()
    }

    private fun setupListeners() {
        binding.btnBack.setOnClickListener {
            navigateBack()
        }

        binding.btnCancel.setOnClickListener {
            navigateBack()
        }

        // Date selection inputs
        binding.edtStartDate.setOnClickListener {
            showDatePicker(binding.edtStartDate)
        }

        binding.edtEndDate.setOnClickListener {
            showDatePicker(binding.edtEndDate)
        }

        // Save campaign
        binding.btnSaveCampaign.setOnClickListener {
            saveCampaign()
        }
    }

    private fun setupInputWatchers() {
        val watcher = object : android.text.TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                updateSaveButtonState()
            }
            override fun afterTextChanged(s: android.text.Editable?) {}
        }

        binding.edtCampaignName.addTextChangedListener(watcher)
        binding.edtCampaignDesc.addTextChangedListener(watcher)
        binding.edtMinOrderValue.addTextChangedListener(watcher)
        binding.edtDiscountValue.addTextChangedListener(watcher)
        binding.edtStartDate.addTextChangedListener(watcher)
        binding.edtEndDate.addTextChangedListener(watcher)
    }

    private fun updateSaveButtonState() {
        val title = binding.edtCampaignName.text.toString().trim()
        val description = binding.edtCampaignDesc.text.toString().trim()
        val minOrder = binding.edtMinOrderValue.text.toString().trim()
        val discount = binding.edtDiscountValue.text.toString().trim()
        val startDate = binding.edtStartDate.text.toString().trim()
        val endDate = binding.edtEndDate.text.toString().trim()

        val allFilled = title.isNotEmpty() && 
                        description.isNotEmpty() && 
                        minOrder.isNotEmpty() && 
                        discount.isNotEmpty() && 
                        startDate.isNotEmpty() && 
                        endDate.isNotEmpty()

        binding.btnSaveCampaign.isEnabled = allFilled
        binding.btnSaveCampaign.alpha = if (allFilled) 1.0f else 0.5f
    }

    private fun showDatePicker(editText: EditText) {
        val calendar = Calendar.getInstance()
        val currentText = editText.text.toString()
        if (currentText.isNotEmpty()) {
            try {
                val date = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).parse(currentText)
                if (date != null) {
                    calendar.time = date
                }
            } catch (e: Exception) {
                // Ignore parsing errors and use current calendar date
            }
        }

        val datePickerDialog = DatePickerDialog(
            requireContext(),
            { _, year, month, dayOfMonth ->
                val selectedCal = Calendar.getInstance()
                selectedCal.set(Calendar.YEAR, year)
                selectedCal.set(Calendar.MONTH, month)
                selectedCal.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                val sdf = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
                editText.setText(sdf.format(selectedCal.time))
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )
        datePickerDialog.show()
    }

    private fun loadCategories() {
        lifecycleScope.launch {
            val loadedCategories = withContext(Dispatchers.IO) {
                val list = mutableListOf<String>()
                list.add("Tất cả sản phẩm") // Default option
                try {
                    val inputStream = requireContext().assets.open("categories.json")
                    val reader = BufferedReader(InputStreamReader(inputStream))
                    val jsonString = reader.use { it.readText() }
                    val jsonObject = JSONObject(jsonString)
                    val jsonArray = jsonObject.getJSONArray("categories")
                    for (i in 0 until jsonArray.length()) {
                        val catObj = jsonArray.getJSONObject(i)
                        val name = catObj.optString("name", "")
                        if (name.isNotEmpty() && !list.contains(name)) {
                            list.add(name)
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                list
            }

            categoriesList.clear()
            categoriesList.addAll(loadedCategories)

            val adapter = ArrayAdapter(
                requireContext(),
                android.R.layout.simple_spinner_item,
                categoriesList
            )
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            binding.spnApplicableProducts.adapter = adapter
        }
    }

    private fun setupOfferTypeSpinner() {
        val adapter = ArrayAdapter(
            requireContext(),
            android.R.layout.simple_spinner_item,
            offerTypes
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spnOfferType.adapter = adapter

        binding.spnOfferType.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                if (position == 0) {
                    binding.txtDiscountLabel.text = "Giá trị (%)"
                    binding.edtDiscountValue.hint = "10"
                } else {
                    binding.txtDiscountLabel.text = "Giá trị (VNĐ)"
                    binding.edtDiscountValue.hint = "50000"
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun saveCampaign() {
        val title = binding.edtCampaignName.text.toString().trim()
        val description = binding.edtCampaignDesc.text.toString().trim()
        val minOrderStr = binding.edtMinOrderValue.text.toString().trim()
        val discountValStr = binding.edtDiscountValue.text.toString().trim()
        val startDateStr = binding.edtStartDate.text.toString().trim()
        val endDateStr = binding.edtEndDate.text.toString().trim()

        if (title.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng nhập tên chương trình!", Toast.LENGTH_SHORT).show()
            return
        }

        if (description.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng nhập mô tả chương trình!", Toast.LENGTH_SHORT).show()
            return
        }

        val minOrderValue = minOrderStr.toLongOrNull() ?: 0L
        val discountValue = discountValStr.toLongOrNull()

        if (discountValue == null || discountValue <= 0) {
            Toast.makeText(requireContext(), "Vui lòng nhập giá trị khuyến mãi hợp lệ!", Toast.LENGTH_SHORT).show()
            return
        }

        if (endDateStr.isEmpty()) {
            Toast.makeText(requireContext(), "Vui lòng chọn ngày kết thúc!", Toast.LENGTH_SHORT).show()
            return
        }

        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                try {
                    // 1. Get writeable file in internal storage
                    val internalFile = File(requireContext().filesDir, "vouchers.json")
                    
                    // 2. Load existing vouchers
                    val jsonArray = if (internalFile.exists()) {
                        val text = internalFile.readText()
                        JSONArray(text)
                    } else {
                        // Load from assets initially
                        val inputStream = requireContext().assets.open("vouchers.json")
                        val text = inputStream.bufferedReader().use { it.readText() }
                        JSONArray(text)
                    }

                    // 3. Generate unique ID & Code
                    var maxVal = 0
                    for (i in 0 until jsonArray.length()) {
                        val item = jsonArray.getJSONObject(i)
                        val idStr = item.optString("id", "")
                        if (idStr.startsWith("v")) {
                            idStr.substring(1).toIntOrNull()?.let { num ->
                                if (num > maxVal) maxVal = num
                            }
                        }
                    }
                    val nextIdNum = maxVal + 1
                    val newId = "v${String.format(Locale.getDefault(), "%03d", nextIdNum)}"

                    // Voucher Type Mapping
                    val voucherType = if (binding.radTypeDiscount.isChecked) "discount" else "free ship"

                    // Offer Type Mapping
                    val offerType = if (binding.spnOfferType.selectedItemPosition == 0) "percentage" else "fixed_amount"

                    // Auto-generated Voucher Code
                    val autoCode = generateVoucherCode(title, offerType, discountValue)

                    // Format HSD: end date with 23:59:59 time
                    val hsdVal = try {
                        val inputFormat = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
                        val date = inputFormat.parse(endDateStr)
                        if (date != null) {
                            val outputFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
                            "${outputFormat.format(date)} 23:59:59"
                        } else {
                            "2026-12-31 23:59:59"
                        }
                    } catch (e: Exception) {
                        "2026-12-31 23:59:59"
                    }

                    // Selected applicable category name
                    val selectedCategory = binding.spnApplicableProducts.selectedItem?.toString() ?: "Tất cả sản phẩm"

                    // 4. Create new voucher JSON object
                    val newVoucher = JSONObject().apply {
                        put("id", newId)
                        put("title", title)
                        put("description", description)
                        put("code", autoCode)
                        put("status", "valid")
                        put("hsd", hsdVal)
                        put("type", voucherType)
                        put("from-gift", false)
                        put("quantity", 50) // Default quantity
                        put("minOrderValue", minOrderValue)
                        put("applicableProducts", selectedCategory)
                        put("offerType", offerType)
                        put("discountValue", discountValue)
                    }

                    // 5. Append & Save
                    jsonArray.put(newVoucher)
                    FileOutputStream(internalFile).use { fos ->
                        fos.write(jsonArray.toString(2).toByteArray())
                    }

                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            Toast.makeText(requireContext(), "Tạo chương trình khuyến mãi thành công!", Toast.LENGTH_SHORT).show()
            navigateBack()
        }
    }

    private fun generateVoucherCode(campaignName: String, offerType: String, discountValue: Long): String {
        // Create a prefix from campaign name
        val words = campaignName.split(" ").filter { it.isNotEmpty() }
        val prefix = words.take(3)
            .joinToString("") { it.take(1).uppercase(Locale.getDefault()) }
            .filter { it.isLetter() }

        val cleanPrefix = if (prefix.isEmpty()) "KM" else prefix
        val suffix = if (offerType == "percentage") "${discountValue}PCT" else "${discountValue / 1000}K"
        val charPool = ('A'..'Z') + ('0'..'9')
        val randomPart = (1..3).map { charPool.random() }.joinToString("")
        
        return "$cleanPrefix$suffix$randomPart".take(12).uppercase(Locale.getDefault())
    }

    private fun navigateBack() {
        val mainActivity = activity as? MainActivity ?: return
        mainActivity.loadFragment(CustomerAdminFragment())
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
